src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -271,89 +275,77 @@ public void AddToPending(ChainedHeaderBlock chainedHeaderBlock)
         {
             this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock.ChainedHeader);
 
-            this.blocksQueue.Enqueue(chainedHeaderBlock);
+            lock (this.batchLock)
+            {
+                this.batch.Add(chainedHeaderBlock);
+                this.currentBatchSizeBytes += chainedHeaderBlock.Block.BlockSize.Value;
+                this.addEvent.Set();
+            }
 
             this.logger.LogTrace("(-)");
         }
 
         /// <summary>
-        /// Dequeues the blocks continuously and saves them to the database when max batch size is reached or timer ran out.
+        /// Periodically checks whether the batch of blocks is full enough to be saved to the database,
+        /// or if sufficient time elapses save the batch anyway.
         /// </summary>
         /// <remarks>Batch is always saved on shutdown.</remarks>
-        private async Task DequeueBlocksContinuouslyAsync()
+        private async Task CheckBatchContinuouslyAsync()
         {
             this.logger.LogTrace("()");
 
-            Task<ChainedHeaderBlock> dequeueTask = null;
-            Task timerTask = null;
-
             while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
             {
-                // Start new dequeue task if not started already.
-                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
-
-                // Wait for one of the tasks: dequeue or timer (if available) to finish.
-                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
-
                 bool saveBatch = false;
 
                 try
                 {
-                    await task.ConfigureAwait(false);
-                }
-                catch (OperationCanceledException)
-                {
-                    // Happens when node is shutting down or Dispose() is called.
-                    // We want to save whatever is in the batch before exiting the loop.
-                    saveBatch = true;
-
-                    this.logger.LogDebug("Node is shutting down. Save batch.");
-                }
-
-                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip
-                // or the max batch size is reached or the node is shutting down.
-                if (dequeueTask.Status == TaskStatus.RanToCompletion)
-                {
-                    ChainedHeaderBlock item = dequeueTask.Result;
-
-                    // Set the dequeue task to null so it can be assigned on the next iteration.
-                    dequeueTask = null;
-
-                    lock (this.getBlockLock)
+                    using (CancellationTokenSource timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.nodeLifetime.ApplicationStopping))
```

Comment:
Wont this be much simpler by just using an the AsyncLoop class and iterate every second on the batch condition?
You guys are complicating things unnecessarily.